### PR TITLE
feat(FQDN): serialization format rpc_host_port

### PR DESCRIPTION
**Pull request by GehaFearless**
issue: https://github.com/apache/incubator-pegasus/issues/1495

Implement thrift protocol functon read and write for class rpc_host_port.
Both about binary protocol and json.

### CODE REVIEW

**Comment by acelyc111** [comment]
@GehaFearless Good! Could you please add some tests to verify that serialize and deserialize both json and binary format is working?

**Review by foreverneverer — DISMISSED** [review]

File: src/runtime/test/host_port_test.cpp [hunk 1]
```diff
@@ -202,4 +211,29 @@ TEST(host_port_test, dns_resolver)
...
+    host_port hp = host_port("localhost", 8080);
+    auto hp_str = hp.to_string();
+    ::dsn::rpc_address server("localhost", 20101);
+
+    dsn::message_ptr mesg_ptr = dsn::message_ex::create_request(RPC_TEST_THRIFT_HOST_PORT_PARSER);
+    mesg_ptr->header->context.u.serialize_format = t;
+
+    ::dsn::marshall(mesg_ptr.get(), hp);
+
+    dsn::task_tracker _tracker;
```
Comment (acelyc111) [inline src/runtime/test/host_port_test.cpp:225]: rename `_tracker` to `tracker`.

File: src/runtime/test/host_port_test.cpp [hunk 2]
```diff
@@ -202,4 +211,29 @@ TEST(host_port_test, dns_resolver)
     }
 }
 
+void send_and_check_host_port_by_different_serialize(dsn_msg_serialize_format t)
+{
+    host_port hp = host_port("localhost", 8080);
+    auto hp_str = hp.to_string();
+    ::dsn::rpc_address server("localhost", 20101);
+
+    dsn::message_ptr mesg_ptr = dsn::message_ex::create_request(RPC_TEST_THRIFT_HOST_PORT_PARSER);
```
Comment (acelyc111) [inline src/runtime/test/host_port_test.cpp:220]: how about msg_ptr? it's a common practice to shorten "message" as "msg".

File: src/runtime/test/host_port_test.cpp [hunk 3]
```diff
@@ -202,4 +211,29 @@ TEST(host_port_test, dns_resolver)
...
+    mesg_ptr->header->context.u.serialize_format = t;
+
+    ::dsn::marshall(mesg_ptr.get(), hp);
+
+    dsn::task_tracker _tracker;
+    rpc::call(server, mesg_ptr.get(), &_tracker, [hp_str](error_code ec, std::string &&resp) {
+        if (ERR_OK == ec) {
```
Comment (acelyc111) [inline src/runtime/test/host_port_test.cpp:227]: Is it always ERR_OK? add ASSERT_EQ(ERR_OK, ec) if it is.

File: src/common/serialization_helper/thrift_helper.h [hunk 4]
```diff
@@ -280,6 +281,105 @@ inline uint32_t rpc_address::write(apache::thrift::protocol::TProtocol *oprot) c
...
+            xfer += iprot->readFieldEnd();
+        }
+        xfer += iprot->readStructEnd();
+    }
+
+    _host = host;
+    _port = static_cast<uint16_t>(port);
+    _type = static_cast<dsn_host_type_t>(type_enum_number);
+    CHECK(_type == HOST_TYPE_INVALID || _type == HOST_TYPE_IPV4,
+          "only invalid or ipv4 can be serialized.");
```
Comment (acelyc111) [inline src/common/serialization_helper/thrift_helper.h:344]:
```suggestion
          "only invalid or ipv4 can be deserialized.");
```

File: src/common/serialization_helper/thrift_helper.h [hunk 5]
```diff
@@ -280,6 +281,105 @@ inline uint32_t rpc_address::write(apache::thrift::protocol::TProtocol *oprot) c
...
+    if (binary_proto != nullptr) {
+        // the protocol is binary protocol
+        xfer += oprot->writeString(_host);
+        xfer += oprot->writeI16(static_cast<int16_t>(_port));
+        xfer += oprot->writeByte(static_cast<int8_t>(_type));
+    } else {
+        // the protocol is json protocol
+        xfer += oprot->writeStructBegin("host_port");
+
+        xfer += oprot->writeFieldBegin("_host", ::apache::thrift::protocol::T_STRING, 1);
```
Comment (acelyc111) [inline src/common/serialization_helper/thrift_helper.h:364]:
```suggestion
        xfer += oprot->writeFieldBegin("host", ::apache::thrift::protocol::T_STRING, 1);
```

File: src/common/serialization_helper/thrift_helper.h [hunk 6]
```diff
@@ -280,6 +281,105 @@ inline uint32_t rpc_address::write(apache::thrift::protocol::TProtocol *oprot) c
...
+        xfer += oprot->writeString(_host);
+        xfer += oprot->writeFieldEnd();
+
+        xfer += oprot->writeFieldBegin("_port", ::apache::thrift::protocol::T_I16, 2);
```
Comment (acelyc111) [inline src/common/serialization_helper/thrift_helper.h:368]:
```suggestion
        xfer += oprot->writeFieldBegin("port", ::apache::thrift::protocol::T_I16, 2);
```

File: src/common/serialization_helper/thrift_helper.h [hunk 7]
```diff
@@ -280,6 +281,105 @@ inline uint32_t rpc_address::write(apache::thrift::protocol::TProtocol *oprot) c
...
+        xfer += oprot->writeI16(static_cast<int16_t>(_port));
+        xfer += oprot->writeFieldEnd();
+
+        xfer += oprot->writeFieldBegin("_type", ::apache::thrift::protocol::T_BYTE, 3);
```
Comment (acelyc111) [inline src/common/serialization_helper/thrift_helper.h:372]:
```suggestion
        xfer += oprot->writeFieldBegin("type", ::apache::thrift::protocol::T_BYTE, 3);
```

File: src/runtime/test/host_port_test.cpp [hunk 8]
```diff
@@ -202,4 +211,29 @@ TEST(host_port_test, dns_resolver)
     }
 }
 
+void send_and_check_host_port_by_different_serialize(dsn_msg_serialize_format t)
+{
+    host_port hp = host_port("localhost", 8080);
```
Comment (acelyc111) [inline src/runtime/test/host_port_test.cpp:216]: How about make `hp` as a parameter, and then add more tests?

**Review by foreverneverer — APPROVED** [review]

**Pull request merged**